src/libraries/System.Private.CoreLib/src/System/Array.Enumerators.cs

```diff
@@ -133,4 +140,39 @@ public T Current
 
         object? IEnumerator.Current => Current;
     }
+
+    internal abstract class GenericEmptyArrayEnumeratorBase
+    {
+#pragma warning disable CA1822 // https://github.com/dotnet/roslyn-analyzers/issues/5911
+        public bool MoveNext() => false;
+
+        public void Reset() { }
+
+        public void Dispose() { }
+#pragma warning restore CA1822
+    }
+
+    /// <summary>Provides an empty enumerator singleton.</summary>
+    /// <remarks>
+    /// If the consumer is using SZGenericArrayEnumerator elsewhere or is otherwise likely
+    /// to be using T[] elsewhere, SZGenericArrayEnumerator's singleton should be used.  Otherwise,
+    /// this singleton should be used, as it doesn't reference T[] in order to reduce footprint.
+    /// </remarks>
+    internal sealed class GenericEmptyArrayEnumerator<T> : GenericEmptyArrayEnumeratorBase, IEnumerator<T>
+    {
+        public static readonly GenericEmptyArrayEnumerator<T> Instance = new();
+
+        private GenericEmptyArrayEnumerator() { }
+
+        public T Current
+        {
+            get
+            {
+                ThrowHelper.ThrowInvalidOperationException_EnumCurrent(-1);
+                return default;
+            }
+        }
+
+        object? IEnumerator.Current => Current;
```

Comment:
This can be in the non-generic base class as well.

Reply:
(Change it to `ThrowHelper.ThrowInvalidOperationException_EnumCurrent(-1)` directly instead of delegating to generic `Current`.)

Reply:
Done